File: examples/node/multi-index.ts
/**
 * Example: Query multiple TreeDex indexes simultaneously (Node.js)
 *
 * Demonstrates building separate indexes for multiple documents and querying
 * them all at once with TreeDex.queryAll().
 */

import { TreeDex, GeminiLLM } from "../../src/index.js";

async function main() {
  const llm = new GeminiLLM(process.env.GEMINI_API_KEY!);

  // --- Build indexes for multiple documents ---
  console.log("Building indexes...");
  const indexA = await TreeDex.fromFile("manual_a.pdf", llm);
  const indexB = await TreeDex.fromFile("manual_b.pdf", llm);
  const indexC = await TreeDex.fromFile("manual_c.pdf", llm);

  // Optionally save and reload later
  await indexA.save("manual_a.json");
  await indexB.save("manual_b.json");
  await indexC.save("manual_c.json");

  // --- Query all indexes at once ---
  const question = "What are the safety guidelines?";

  console.log(`\nQuerying all indexes: '${question}'`);
  const multi = await TreeDex.queryAll(
    [indexA, indexB, indexC],
    question,
    { labels: ["Manual A", "Manual B", "Manual C"] },
  );

  // Inspect per-index results
  multi.results.forEach((result, i) => {
    console.log(`\n[${multi.labels[i]}]`);
    console.log(`  Nodes:  ${JSON.stringify(result.nodeIds)}`);
    console.log(`  Pages:  ${result.pagesStr}`);
    console.log(`  Reason: ${result.reasoning}`);
  });

  // Combined context with [Manual A] / [Manual B] / [Manual C] headers
  console.log("\n--- Combined Context ---");
  console.log(multi.combinedContext.slice(0, 500), "...");

  // --- Agentic mode: one answer across all documents ---
  console.log("\nQuerying with agentic answer...");
  const multiAgentic = await TreeDex.queryAll(
    [indexA, indexB, indexC],
    question,
    { labels: ["Manual A", "Manual B", "Manual C"], agentic: true },
  );
  console.log(`\nAnswer:\n${multiAgentic.answer}`);
}

main().catch(console.error);

File: examples/multi_index.py
"""Example: Query multiple TreeDex indexes simultaneously.

Demonstrates building separate indexes for multiple documents and querying
them all at once with TreeDex.query_all().
"""

import os

from treedex import TreeDex, GeminiLLM


def main():
    llm = GeminiLLM(api_key=os.environ["GEMINI_API_KEY"])

    # --- Build indexes for multiple documents ---
    print("Building indexes...")
    index_a = TreeDex.from_file("manual_a.pdf", llm=llm)
    index_b = TreeDex.from_file("manual_b.pdf", llm=llm)
    index_c = TreeDex.from_file("manual_c.pdf", llm=llm)

    # Optionally save and reload later
    index_a.save("manual_a.json")
    index_b.save("manual_b.json")
    index_c.save("manual_c.json")

    # --- Query all indexes at once ---
    question = "What are the safety guidelines?"

    print(f"\nQuerying all indexes: '{question}'")
    multi = TreeDex.query_all(
        [index_a, index_b, index_c],
        question,
        labels=["Manual A", "Manual B", "Manual C"],
    )

    # Inspect per-index results
    for i, result in enumerate(multi.results):
        print(f"\n[{multi.labels[i]}]")
        print(f"  Nodes:   {result.node_ids}")
        print(f"  Pages:   {result.pages_str}")
        print(f"  Reason:  {result.reasoning}")

    # Combined context with [Manual A] / [Manual B] / [Manual C] headers
    print("\n--- Combined Context ---")
    print(multi.combined_context[:500], "...")

    # --- Agentic mode: one answer across all documents ---
    print("\nQuerying with agentic answer...")
    multi_agentic = TreeDex.query_all(
        [index_a, index_b, index_c],
        question,
        labels=["Manual A", "Manual B", "Manual C"],
        agentic=True,
    )
    print(f"\nAnswer:\n{multi_agentic.answer}")


if __name__ == "__main__":
    main()
